File: client/assets/tools/on.js
/* TODO
Specific unuse in component also
*/

export const on = (target, { run, ...options } = {}) => {
  return new Proxy(() => {}, {
    get(_, key) {
      const type = key;
      return new Proxy(() => {}, {
        get(_, key) {
          return (...args) => {
            const handler = Handler(run, args);
            if (key === "use") {
              target.addEventListener(type, handler, options);
              return handler;
            }
            if (key === "unuse") {
              target.removeEventListener(type, handler, options);
              return target;
            }
            throw new Error(`Invalid key: ${key}`);
          };
        },
        apply(_, thisArg, args) {
          const handler = Handler(run, args);
          target.addEventListener(type, handler, options);
          return handler;
        },
      });
    },
    set(_, type, handler) {
      if (run) {
        handler();
      }
      target.addEventListener(type, handler, options);
      return true;
    },
  });
};

function Handler(run, args) {
  const handler = args.find((a) => typeof a === "function");
  if (run) {
    handler();
  }
  return handler;
}

File: _history/server/client_code/20260310t1600/tools/_on.py
class on:
    def __init__(self, target, *args, **options):
        self.target = target
        self.args = args
        self.options = options

    def __call__(self, handler: callable) -> callable:
        _handler = handler() if isinstance(handler, type) else handler
        type_ = self.args[0] if self.args else handler.__name__
        self.target.addEventListener(type_, _handler, self.options)
        return handler
